core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java

```diff
@@ -299,39 +303,49 @@ public void runScenarioTests(TestPlan testPlan, DeploymentCreationResult deploym
                 }
             }
 
-
-            for (TestScenario testScenario : testPlan.getTestScenarios()) {
-                try {
-                    scenarioExecutor.execute(testScenario, deploymentCreationResult, testPlan);
-                    Optional<ResultParser> parser = ResultParserFactory.getParser(testPlan, testScenario);
-                    if (parser.isPresent()) {
-                        try {
-                            ResultParser resultParser = parser.get();
-                            resultParser.parseResults();
-                            resultParser.archiveResults();
-                        } catch (ResultParserException e) {
-                            logger.error(String.format("Error parsing the results for the" +
-                                    " scenario %s", testScenario.getName()));
+            List<ConfigChangeSet> configChangeSetList = testPlan.getScenarioConfig().getConfigChangeSets();
+            // Run test with config change sets
+            if (configChangeSetList != null) {
+                OSCategory osCategory = getOSCatagory(testPlan.getInfraParameters());
+                Optional<ConfigChangeSetExecutor> configChangeSetExecutor =
+                        ConfigChangeSetFactory.getExecuter(osCategory);
+                if (configChangeSetExecutor.isPresent()) {
+                    // Initialize config set repos on agent
+                    if (configChangeSetExecutor.get().initConfigChangeSet(testPlan)) {
+                        for (ConfigChangeSet configChangeSet : configChangeSetList) {
+                            logger.info("Start running config change set for " + configChangeSet.getName());
+                            // Apply config change set script on agent
+                            if (configChangeSetExecutor.get().applyConfigChangeSet(testPlan, configChangeSet,
+                                            deploymentCreationResult, true)) {
+                                // Run test scenarios for relevant config change set
+                                for (TestScenario testScenario : testPlan.getTestScenarios()) {
+                                    if (configChangeSet.getName().equals(testScenario.getConfigChangeSetName())) {
+                                        executeTestScenario(testScenario, deploymentCreationResult, testPlan);
+                                    }
+                                }
+                                // Revert config change set after running test scenarios
+                                // If revert back is not success then, break all test scenarios running
+                                if (!configChangeSetExecutor.get().applyConfigChangeSet(testPlan, configChangeSet,
```

Comment:
IMO, we need to have two separate methods to handle #apply and #revert scenarios.